File: frontend/src/apis/problem/ListProblemDataFiles.ts
import type { APIWithSessionMessage } from '@/system/api/api-message'
import type { ContestSlug } from '@/objects/contest/ContestSlug'
import { contestSlugValue } from '@/objects/contest/ContestSlug'
import type { ProblemDataFileListResponse } from '@/objects/problem/response/ProblemDataFileListResponse'
import type { ProblemSlug } from '@/objects/problem/ProblemSlug'
import { problemSlugValue } from '@/objects/problem/ProblemSlug'

/** 列出题目数据文件名；输入题目 slug 和可选比赛上下文，输出文件名集合。 */
export class ListProblemDataFiles implements APIWithSessionMessage<ProblemDataFileListResponse> {
  declare readonly responseType?: ProblemDataFileListResponse
  readonly method = 'GET'
  readonly apiPath: string

  constructor(problemSlug: ProblemSlug, contestSlug?: ContestSlug) {
    const params = new URLSearchParams()
    if (contestSlug) {
      params.set('contestSlug', contestSlugValue(contestSlug))
    }
    this.apiPath = `problems/${problemSlugValue(problemSlug)}/data/files${params.size > 0 ? `?${params.toString()}` : ''}`
  }

  body(): undefined {
    return undefined
  }
}

File: backend/src/main/scala/domains/problem/api/ListProblemDataFiles.scala
package domains.problem.api

import cats.effect.IO
import domains.auth.api.AuthenticatedApi
import domains.auth.objects.internal.AuthenticatedUser

import domains.problem.objects.response.ProblemDataFileListResponse
import io.circe.Encoder
import org.http4s.{Method, Request, Status}
import shared.api.{ApiPath, PathParams}

import java.sql.Connection

/** 列出题目数据文件名的管理端 API；从对象存储读取当前文件列表，要求题目管理权限。 */
final case class ListProblemDataFiles(problemDataStorage: ProblemDataStorageContext)
    extends AuthenticatedApi[ProblemManagementContext, ProblemDataFileListResponse]:

  override val method: Method = Method.GET
  override val path: ApiPath = ApiPath("/api/problems/:problemSlug/data/files")
  override val successStatus: Status = Status.Ok
  override protected val outputEncoder: Encoder[ProblemDataFileListResponse] = summon[Encoder[ProblemDataFileListResponse]]

  /** 解析题目管理上下文。 */
  override def decode(request: Request[IO], pathParams: PathParams): IO[ProblemManagementContext] =
    ProblemManagementContext.decode(request, pathParams)

  /** 校验管理权限后列出对象存储中的文件名；目录路径会在存储扩展方法中折叠为文件名。 */
  override def plan(
    connection: Connection,
    actor: AuthenticatedUser,
    context: ProblemManagementContext
  ): IO[ProblemDataFileListResponse] =
    ProblemManagementContext
      .requireManagedProblem(connection, actor, context)
      .flatMap(problem => ProblemDataStorage.listFiles(problemDataStorage, problem.slug).map(ProblemDataFileListResponse(_)))
